core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -94,8 +88,23 @@ class LogSegment private[log] (val log: FileRecords,
   private var rollingBasedTimestamp: Option[Long] = None
 
   /* The maximum timestamp we see so far */
-  @volatile private var maxTimestampSoFar: Long = timeIndex.lastEntry.timestamp
-  @volatile private var offsetOfMaxTimestamp: Long = timeIndex.lastEntry.offset
+  @volatile private var _maxTimestampSoFar: Long = -1L
```

Comment:
Should we use an `Option` instead?

Reply:
I don't think `Option` can provide the atomicity guarantee without using a lock, so I think it is safer to just keep `volatile` Long as before.